### DISCUSSION TITLE
Code blocks not rendering properly

### DISCUSSION THREAD

**Discussion by nicoalba** [squidfunk/mkdocs-material#2490] [Q&A]
Hello,

New to this world so forgive my ignorance. Wasn't sure if this should go in discussion or issues but I figure it's an operator problem so I decided to post here. I'm using MkDocs w/ Material theme for documentation. Love the theme but code blocks aren't rendering properly for me even though they render just fine in my Visual Studio MD preview (and in other themes) and also in my GitLab preview.

My code:
![gh1](https://user-images.githubusercontent.com/49991668/112521739-7ab62180-8d62-11eb-9f46-cbef1ce19fe0.png)

Visual Studio MD preview: 
![gh2](https://user-images.githubusercontent.com/49991668/112521782-86094d00-8d62-11eb-9895-db4bf56f4789.png)

MkDocs result:
![gh3](https://user-images.githubusercontent.com/49991668/112522277-0f208400-8d63-11eb-9642-e98eecdd326e.PNG)

I have messed around with indentations and spacing and all kinds of things. Any help?

**Comment by facelessuser** [comment]
Please provide your mkdocs config as I think you are missing the SuperFences plugin. Fenced code blocks are not part of the original Markdown spec, so they are not enabled by default.

**Reply by nicoalba** [reply]
Hi @facelessuser , thanks for the reply. Here's my mkdocs config file:

![snip1](https://user-images.githubusercontent.com/49991668/112547412-55381080-8d80-11eb-883e-0b698ed39bc6.PNG)

I don't think I have the SuperFences plugin.

**Reply by facelessuser** [reply]
I can't tell much from the screenshot, but here is info to get fenced code blocks working: https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#highlight.

**Reply by nicoalba** [reply]
@facelessuser Hi, I just went through the instructions and I already have it installed. Both Pymdown-extensions and Pygment.  Let me read through your link and see if there's something I need to add to configuration.

The config file is just very long but I can zip and attach if that helps.

**Reply by facelessuser** [reply]
You would need to just provide a code short here of just the extension setup. I can't tell anything right now.

**Reply by nicoalba** [reply]
@facelessuser Thanks so much. The problem was I didn't even have a configuration section in my mkdocs file. I just had site nav and some general info up top. I added:

markdown_extensions:
- pymdownx.highlight
- pymdownx.superences

and now it works :D. Like I said, I'm a noob. Thanks!

**Reply by facelessuser** [reply]
That was the missing info. Glad you figured it out 😊.

**Reply by nicoalba** [reply]
@facelessuser Would you happen to know about making the code blocks collapsible? Can you point me to any instructions? I don't see the option here: https://squidfunk.github.io/mkdocs-material/reference/code-blocks/

**Reply by facelessuser** [reply]
Collapsible? I'm not sure what you mean. I'm pretty sure that whatever you are referring to, it would require additional CSS.

**Reply by squidfunk** [reply]
You could wrap your code blocks in details ([collapsible admonitions](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#collapsible-blocks))

**Reply by facelessuser** [reply]
> You could wrap your code blocks in details ([collapsible admonitions](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#collapsible-blocks))

Ah...why didn't I put that together 🤦🏻 .

**Reply by squidfunk** [reply]
@facelessuser still thinking about adding a special admonition to allow code blocks with titles, similar to the `only-child` hack on content tabs 😉

**Reply by facelessuser** [reply]
You could probably pull it off with custom fences as well. I do special fences with them, like for instance these that add custom color gradients or color swatches from the code block as a header: https://facelessuser.github.io/coloraide/interpolation/#interpolating.

TBH, a lot more than just that is happening, but you get the idea.

**Reply by facelessuser** [reply]
If I recall, you could probably override all the default fences with a specialized code block format that does titles. I'm pretty sure I allow overriding the base default.

**Reply by facelessuser** [reply]
You could probably do this even easier by attaching an attribute with a title and just using some CSS to add a simple title bar using that attribute's content. Of course, I would have to allow arbitrary data attributes to be attached to code blocks...

**Reply by squidfunk** [reply]
Thanks for dumping some ideas! The data attribute approach is what’s been on my mind for some time, but currently not possible due to the limitations. I guess stealing `data-*` attributes before pygments eats them all up could unlock a new level of flexibility.

**Reply by squidfunk** [reply]
I guess at some point we need to write a spec on all the Markdown extensions you/we added with PymdownX and Material for MkDocs 😉 it’s far beyond the original spec and proven to be very workable.

**Reply by facelessuser** [reply]
Yup, `data-*` has been on my mind. I need to play with it sometime soon.

**Reply by nicoalba** [reply]
I'm glad my newbness was able to spur a fruitful convo :D

**Comment by matchavez** [comment]
As requested by Martin to put here, I would find it helpful if there was a way that a code block - just the basic code block - would return the language as it is marked. I particularly like the way Typora handles this, so I'll put an example here, in images.
<img width="594" alt="Screen Shot 2021-06-17 at 20 45 32" src="https://user-images.githubusercontent.com/1787040/122363460-0e3b6f00-cfad-11eb-80ee-61356959d292.png"><img width="932" alt="Screen Shot 2021-06-17 at 20 45 58" src="https://user-images.githubusercontent.com/1787040/122363452-0c71ab80-cfad-11eb-9d43-d7d17d40ae92.png">
It should be noted that this might not be the best way, but any identification of the language would be better than having to do a super-fences workaround. I know this is somewhat discussed already, but I wanted to share a "living" example. 1,000 ways to do it visually, but of course programmatically is the tricky part, and I thank you for looking at it. 🥇

**Reply by jaceklaskowski** [reply]
I'm curious why it's important? What would that language indication bring to users? I've got no opinion on this and am simply curious what I'm missing.

**Reply by matchavez** [reply]
Well, the main reason is that as the writer of the document, you have an option to put in the language after the three backticks. There's no reason that should be hidden. If I indicate something is written, and therefore highlighted based on a language, that should not be hidden away in the markdown. Every bit of text in Markdown is used… if you put something that formats, it formats. If you write content, it displays content. This is the only thing that's content in markdown, but doesn't get rendered through by design. Specifically, you don't want to be constrained to having to do a tab structure just to point out what the language is; it's a clunky workaround.

Even if it were a hover, or something similar, that's fine. But if you identify a block, that should get rendered.

**Reply by facelessuser** [reply]
TBH, this is a behavior of Pygments. Pygments, once it is highlighted the text, doesn't need the language class in the final HTML output as it has already processed the code block. This is not true for JavaScript parsers that run on the final HTML. So Pygments does not inject the `language-<code>` as a class in the final output.

If anything, such requests should be presented upstream in relevant packages. For instance, if you use SuperFences (which I am the author of) it is possible that the class could be injected there. To some extent, I'd argue this to be a request for Pygments, but I've seen this request multiple times, and have been considering making something available that indicates what language the code block was specified as (whether a class of `language-<code>` is applied or some `data-language` attribute is yet to be seen).

So, I don't think this is really something for MkDocs Material to implement, but some some library upstream.

**Comment by matchavez** [comment]
So I understand exactly here, what you're saying is Pygments gets the language-lang class, uses it, and doesn't return it in any way. Given that's the case, how do you handle line numbers and highlighting? Do you read that first? If so, you'd read the lang at that same time, no? I'm sure that once it's off to Pygments, it's too late, but how do you handle `linenums` ?

**Reply by facelessuser** [reply]
Pygments handles line numbers. It has an option. So we (Markdown extensions -- not Mkdocs Material) pass it to Pygments. We do the same with language classes. I'm not saying we cannot intercept the language. I'm also not saying it cannot be injected. What I'm saying is:

- MkDocs Material is a theme and does not control this in any way and will not be the one that implements this desired behavior. If it was implemented, they could do something with the exposed attribute or class, but the project itself would not be the one to expose this attribute in code blocks.
- A library upstream would need to handle this and expose it in the output. It could be Pygments itself, but it could be one of the Markdown extensions that use Pygments, like SuperFences (which I am the author of). It is not impossible to do what you are requesting, but it would have to be done upstream if was to be exposed.

Anyways, hopefully, this better summarizes what I'm trying to convey.

**Reply by matchavez** [reply]
I follow you. I'll take it up with Pygments, and you guys can make it look pretty if they do it. :)

**Comment by elissasoroj** [comment]
Hello! 
I'm having the same issue as the OP - but the fix doesn't seem to be the same. My code blocks aren't rendering properly, but I already have the SuperFences plugin. 

**Code:**
<img width="761" alt="Screen Shot 2022-07-17 at 5 17 06 PM" src="https://user-images.githubusercontent.com/71294603/179425137-e28094cf-70a2-45f7-b858-b98c63e24914.png">

**Render:**
<img width="761" alt="Screen Shot 2022-07-17 at 5 18 44 PM" src="https://user-images.githubusercontent.com/71294603/179425161-3166a76f-1313-4339-9dac-bd359d28b972.png">

**mkdocs.yaml:**
<img width="761" alt="Screen Shot 2022-07-17 at 5 19 29 PM" src="https://user-images.githubusercontent.com/71294603/179425170-aab01ddf-1926-43ff-8d6d-51ac9e8970d4.png">

**Reply by facelessuser** [reply]
What if you put a new line between the header line and the code block?

**Reply by elissasoroj** [reply]
I tried all the different permutations of spaces, tabs, etc. 
Turns out it's some weird issue with the local deploy. When I look at the published site everything renders fine (local deploy still doesn't): 
<img width="798" alt="Screen Shot 2022-07-17 at 7 29 09 PM" src="https://user-images.githubusercontent.com/71294603/179429150-3729bb96-4e74-4bbe-b673-ecf980ae4767.png">

Unfortunately that's not the end of it. On this site the code blocks are displaying, but on my other site they are not, even on the published version (both sites have more or less the same settings in the .yaml):
<img width="769" alt="Screen Shot 2022-07-17 at 7 31 31 PM" src="https://user-images.githubusercontent.com/71294603/179429255-d57d29e3-b9e7-4db8-9775-8f99f72fd0ca.png">

.yml for that site, for good measure:
<img width="638" alt="Screen Shot 2022-07-17 at 7 33 52 PM" src="https://user-images.githubusercontent.com/71294603/179429283-bd5c0d31-d79f-4dc8-92da-4f4d33359ecc.png">

**Reply by squidfunk** [reply]
You probably need to remove `codehilite`, [as it's not supported anymore](https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown/#codehilite).

**Reply by elissasoroj** [reply]
that seems to have done the trick - thank you!
